### PR TITLE
Add validator stake/unstake support

### PR DESCRIPTION
**Pull request by vihu**

### CODE REVIEW

File: src/lwt.erl [hunk 1]
```diff
@@ -61,6 +71,47 @@ handle_info(_Any, State) ->
...
+            NewPendingOps =
+                State#state.pending_operations ++
+                    [{stake_validator, Owner, ValidatorAddress, ValidatorDCCost}],
+            {reply, ok, State#state{
+                holders = NewHolders,
+                validators = NewValidators,
+                pending_operations = NewPendingOps
+            }}
+    end;
+handle_call({unstake_validator, Owner, ValidatorAddress}, _From, State) ->
```
Comment (Vagabond) [inline src/lwt.erl]: One thing you probably need to account for is that unstake is async (i.e. on mainnet it's 250,000 blocks).

File: src/lwt_chain.erl [hunk 2]
```diff
@@ -9,22 +9,29 @@
...
     pending_rewards = #{}
 }).
 
 -define(HotspotAddFee, 100).
+-define(ValidatorAddFee, 10000).
 
 fund_owner(Payer, DCAmt) ->
     gen_server:call(?MODULE, {fund_owner, Payer, DCAmt}, infinity).
 
+stake_validator(Owner, ValidatorAddress) ->
```
Comment (Vagabond) [inline src/lwt_chain.erl]: Why is this exported, this should only come from the l1 chain?

File: src/lwt_chain.erl [hunk 3]
```diff
@@ -80,16 +87,36 @@ handle_info(oracle, State = #state{oracles = Oracles0, pending_rewards = Rewards
...
+                        fun
+                            ({dc, Account, Value}, {HAcc, VAcc}) ->
+                                lager:info("Crediting ~p with ~p", [Account, Value]),
+                                {credit(Account, Value, HAcc), VAcc};
+                            (
+                                {stake_validator, Owner, ValidatorAddress, ValidatorDCCost},
+                                {HAcc, VAcc}
+                            ) ->
+                                {
+                                    debit(Owner, ValidatorDCCost, HAcc),
```
Comment (Vagabond) [inline src/lwt_chain.erl]: staking is not done in DC, so there should be no debiting/crediting going on here.

File: src/lwt.erl [hunk 4]
```diff
@@ -120,25 +225,51 @@ handle_call(
...
 debit(Key, Amount, Map) ->
     maps:update_with(Key, fun(V) -> V - Amount end, Map).
 
 credit(Key, Amount, Map) ->
     maps:update_with(Key, fun(V) -> V + Amount end, Amount, Map).
+
+add_validator(Key, Val, Map) ->
+    maps:put(Key, Val, Map).
+
+remove_validator(Key, Map) ->
```
Comment (jadeallenx) [inline src/lwt.erl]: isn't this the point where you'd want to return whatever stake?

File: src/lwt_chain.erl [hunk 5]
```diff
@@ -8,22 +14,30 @@
...
+
+-export([height/0, state/0]).
 
 -record(state, {
     oracles = [],
     %% account => balance
     dc_balances = #{},
     %% address => owner
     hotspots = #{},
+    %% val_address => {owner, init_height}
```
Comment (jadeallenx) [inline src/lwt_chain.erl]: is overstaking going to be a thing here? should you record the stake in the metadata? (also shouldn't this data structure be defined as a record?)

File: src/lwt_chain.erl [hunk 6]
```diff
@@ -119,3 +196,12 @@ credit(Key, Amount, Map) ->
 
 debit(Key, Amount, Map) ->
     maps:update_with(Key, fun(V) -> V - Amount end, Map).
+
+add_validator(Key, Value, Map) ->
+    maps:put(Key, Value, Map).
+
+remove_validator(Key, Map) ->
```
Comment (jadeallenx) [inline src/lwt_chain.erl]: again, shouldn't you handle a stake return here?

File: src/lwt.erl (hunk 4 above)
Comment (vihu) [reply src/lwt.erl]: Yes it is _but_ I think you're allowed to unstake whenever you want despite the fact that your staking_period may not have been exhausted. We could make that behavior invalid if we want to though. But the more important bit is that even if you unstaked your stake wouldn't be refunded till the staking period has been exhausted.

Furthermore, the only thing that the `lwt` contract can really do is sort of roll up changes to be "at some point" applied to the `lwt_chain`. I do have it working here but the bug is that the stake is getting credit twice so I was trying to figure that out and/or try some other approach.

File: src/lwt_chain.erl (hunk 5 above)
Comment (vihu) [reply src/lwt_chain.erl]: > is overstaking going to be a thing here?

Uh, maybe. Don't really wanna go there in first pass

> should you record the stake in the metadata? (also shouldn't this data structure be defined as a record?)

We could yeah, but more interesting/relevant piece is that LWT should get debited/locked out of owner's account which _is_ tracked in `lwt` contract's state.

File: src/lwt.erl (hunk 4 above)
Comment (jadeallenx) [reply src/lwt.erl]: oh there will be a "cooldown" period here too then? i would suggest putting the cooldown block in the stake metadata then, or maybe track cooldowns in their own datastructure by block height, then when the block height tick executes, you can check that map and see if there's any pending operations

Comment (evanmcc) [reply src/lwt.erl]: if we don't have some sort of periodic tick, we could force the unstaker to wait through the cooldown then issue a finalize txn that's only valid after the cooldown.

**Pull request merged**